### DISCUSSION TITLE
Prevent pymupdf4llm from removing hyphens at the end of lines

### DISCUSSION THREAD

**Discussion by Froschsonnenkoenigin** [pymupdf/PyMuPDF-Utilities#141] [Q&A]
I tried pymupdf4llm for multi-column PDF documents. It did a good job. But in the extracted text the hyphens are missing at the end of the line. But these are needed to remove the paragraphs and connect the word parts that belong together in order to be able to search the text document properly.

What would I have to change in pymupdf4llm so that hyphens are not removed?

**Comment by JorjMcKie** [comment]
Hm, actually, quite some effort has been invested to recognize and resolve hyphenation.
This happened acknowledging that especially searching for keywords **_requires_** to remove word segmentation as caused by hyphenation - as much as possible at least: there are natural limits to this if thinking of page and column breaks.
So I am slightly surprised about how this could **_prevent_** you from successfully finding keywords.

**Reply by JorjMcKie** [reply]
Anyway, you may want to make a change here and remove that replace occurrence:
![image](https://github.com/user-attachments/assets/c5b47f2b-b84d-4e72-b15d-ab13cc937bb6)

**Comment by Froschsonnenkoenigin** [comment]
Sorry, the problem is the other way around. The hyphens are normally removed correctly so that the separated words are joined correctly. However, in the multi-column document, I have experienced that the words at the end of the line that end without a hyphen are merged with the word in the new line instead of a space being added there.

I have replaced the following line without success:
`md_string = md_string.replace(" ,", ",").replace("\n", "")
`
with:
`import re`
`md_string = re.sub(r"-\n", "", md_string)`
`md_string = re.sub(r"(?<!-)\n", " ", md_string)`
`md_string = md_string.replace(" ,", ",")`

But this did not solve the problem. Because then there are spaces where there were hyphens. What can be done?
